src/wallet/scriptpubkeyman.cpp

```diff
@@ -1717,42 +1717,160 @@ std::unordered_set<CScript, SaltedSipHasher> LegacyDataSPKM::GetScriptPubKeys()
-            int wit_ver = -1;
-            std::vector<unsigned char> witprog;
-            if (script.IsWitnessProgram(wit_ver, witprog) && wit_ver == 0) {
+            break;
+        case TxoutType::WITNESS_V0_KEYHASH:
+        {
+            CKeyID key_id{uint160(sols[0])};
+            CPubKey pubkey;
+            if (GetPubKey(key_id, pubkey) && pubkey.IsCompressed() && HaveKey(key_id)) {
                 spks.insert(script);
+                spks.insert(GetScriptForDestination(ScriptHash(script)));
             }
-        } else {
-            // Multisigs are special. They don't show up as ISMINE_SPENDABLE unless they are in a P2SH
-            // So check the P2SH of a multisig to see if we should insert it
-            std::vector<std::vector<unsigned char>> sols;
-            TxoutType type = Solver(script, sols);
-            if (type == TxoutType::MULTISIG) {
-                CScript ms_spk = GetScriptForDestination(ScriptHash(script));
-                if (IsMine(ms_spk) != ISMINE_NO) {
-                    spks.insert(ms_spk);
+            break;
+        }
+        case TxoutType::MULTISIG:
+        {
+            // Multisigs are only spendable if we have all of their keys
+            std::vector<std::vector<unsigned char>> keys(sols.begin() + 1, sols.begin() + sols.size() - 1);
+            if (!HaveKeys(keys, *this)) {
+                break;
+            }
+            // Multisigs are always spendable inside of P2SH scripts
+            spks.insert(GetScriptForDestination(ScriptHash(script)));
+            // We need to have the P2WSH script for the P2WSH to be spendable.
+            // But the keys also must be compressed
+            CScript ms_wsh = GetScriptForDestination(WitnessV0ScriptHash(script));
+            if (HaveCScript(CScriptID(ms_wsh))) {
+                if (all_keys_compressed(keys)) {
+                    spks.insert(ms_wsh);
+                    spks.insert(GetScriptForDestination(ScriptHash(ms_wsh)));
                 }
             }
+            break;
+        }
+        }
+    }
+    // Iterate again for all the P2WSH scripts
+    for (const auto& script_pair : mapScripts) {
+        const CScript& script = script_pair.second;
+        std::vector<std::vector<unsigned char>> sols;
+        TxoutType type = Solver(script, sols);
+        if (type == TxoutType::WITNESS_V0_SCRIPTHASH) {
+            uint160 hash{RIPEMD160(sols[0])};
+            CScript witness_script;
+            int wit_ver = -1;
+            std::vector<unsigned char> wit_prog;
+            if (GetCScript(CScriptID(hash), witness_script) &&
+                !witness_script.IsPayToScriptHash() &&
+                !witness_script.IsWitnessProgram(wit_ver, wit_prog) &&
+                spks.count(witness_script) > 0) {
+                spks.insert(script);
+                spks.insert(GetScriptForDestination(ScriptHash(script)));
+            }
```

Comment:
Is this being tested? It seems we have P2WSH(multisig) tests inside `ismine_tests.cpp` and `wallet_migration.py` but these type of script should be covered inside the first loop already. Because if the wallet owns the P2WSH, it will also contain the inner multisig script, which will appear in the first loop.

Reply:
I've added additional tests to `wallet_migration.py`.